runtime/parachains/src/session_info.rs

```diff
@@ -253,4 +256,26 @@ mod tests {
 			assert_eq!(session.needed_approvals, 42);
 		})
 	}
+
+	#[test]
+	fn session_pruning_avoids_heavy_loop() {
```

Comment:
I can't figure out how this test ensures that we don't call `Sessions::remove` for every `n < start`.

Reply:
My laptop froze when running this test without the fix (on 1046e95).
It calls `SessionInfo::initializer_on_new_session(&notification)` on a very high block session index.

Reply:
Maybe relying on a cargo test timeout is not most elegant way, but this test shouldn't normally fail.

Reply:
Maybe we could add an internal timeout with a few seconds' max time? Like wrap the whole closure in `async_std::task::block_on(async move { ... }.timeout(Duration::from_secs(5))`, using `subsystem_util::TimeoutExt`. It would mean adding async_std as a dev-dependency, but I don't think that should be a huge deal.